gcc/rust/expand/rust-expand-visitor.cc

```diff
@@ -18,38 +18,199 @@
 
 #include "rust-expand-visitor.h"
 #include "rust-attributes.h"
+#include "rust-ast.h"
+#include "rust-type.h"
+#include "rust-derive.h"
 
 namespace Rust {
 
+bool
+is_derive (AST::Attribute &attr)
+{
+  auto path = attr.get_path ();
+  return attr.has_attr_input ()
+	 && attr.get_attr_input ().get_attr_input_type ()
+	      == AST::AttrInput::TOKEN_TREE
+	 && path == "derive";
+}
+
+bool
+is_builtin (AST::Attribute &attr)
+{
+  auto &segments = attr.get_path ().get_segments ();
+  return !segments.empty ()
+	 && !Analysis::BuiltinAttributeMappings::get ()
+	       ->lookup_builtin (segments[0].get_segment_name ())
+	       .is_error ();
+}
+
 /* Expand all of the macro invocations currently contained in a crate */
 void
 ExpandVisitor::go (AST::Crate &crate)
+{
+  expand_inner_items (crate.items);
+}
+
+/**
+ * Returns a list of all the derive macros to apply, as well as the Attribute
+ * they are from.
+ *
+ * ```rust
+ * #[derive(Clone, Copy)] // attr1
+ * struct S;
+ *
+ * // returns [{Clone, &attr1}, {Copy, &attr1}]
+ *
+ * #[derive(Clone)] // attr1
+ * #[derive(Copy, PartialEq, Ord)] // attr2
+ * struct S;
+ *
+ * // returns [{Clone, &attr1}, {Copy, &attr2}, {PartialEq, &attr2}, {Ord,
+ * &attr2}]
+ * ```
+ *
+ * @param outer_attrs The list of attributes on the item to derive
+ */
+static std::vector<
+  std::pair<std::string, std::reference_wrapper<const AST::Attribute>>>
+get_traits_to_derive (std::vector<AST::Attribute> &outer_attrs)
+{
+  std::vector<
+    std::pair<std::string, std::reference_wrapper<const AST::Attribute>>>
+    to_derive;
+  for (auto it = outer_attrs.begin (); it != outer_attrs.end ();)
+    {
+      auto &attr = *it;
+
+      if (is_derive (attr))
+	{
+	  auto &input = attr.get_attr_input ();
+	  switch (input.get_attr_input_type ())
+	    {
+	      // isn't there a better way to do this?? like parse it or
+	      // something idk. some function I'm not thinking of?
+	      case AST::AttrInput::TOKEN_TREE: {
+		auto &tokens = static_cast<AST::DelimTokenTree &> (input)
+				 .get_token_trees ();
+
+		// erase the delimiters
+		rust_assert (tokens.size () >= 3);
+		tokens.erase (tokens.begin ());
+		tokens.pop_back ();
+
+		for (auto &token : tokens)
+		  {
+		    // skip commas, as they are part of the token stream
+		    if (token->as_string () == ",")
+		      continue;
+
+		    to_derive.emplace_back (token->as_string (), attr);
```

Comment:
This should be parsing and name resolving paths (e.g. `#[derive(std::clone::Clone)]`). But that can be raised as a follow-up issue.

Reply:
Yes, that should be done within the Early Name Resolver but I wanted to get the expansion part done first. This will need to be done before #2172 can be closed (as well as the procedural macros project)